File: sample_backbone_app/src/main/java/org/researchstack/backboneapp/MainActivity.java
package org.researchstack.backboneapp;

import android.content.Intent;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.Toolbar;
import android.util.Base64;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import org.researchstack.backbone.StorageAccess;
import org.researchstack.backbone.answerformat.AnswerFormat;
import org.researchstack.backbone.answerformat.BooleanAnswerFormat;
import org.researchstack.backbone.answerformat.ChoiceAnswerFormat;
import org.researchstack.backbone.answerformat.DateAnswerFormat;
import org.researchstack.backbone.answerformat.IntegerAnswerFormat;
import org.researchstack.backbone.answerformat.TextAnswerFormat;
import org.researchstack.backbone.answerformat.UnknownAnswerFormat;
import org.researchstack.backbone.interop.ViewBackboneInteropTaskActivity;
import org.researchstack.backbone.model.Choice;
import org.researchstack.backbone.model.ConsentDocument;
import org.researchstack.backbone.model.ConsentSection;
import org.researchstack.backbone.model.ConsentSignature;
import org.researchstack.backbone.result.StepResult;
import org.researchstack.backbone.result.TaskResult;
import org.researchstack.backbone.step.ConsentDocumentStep;
import org.researchstack.backbone.step.ConsentSignatureStep;
import org.researchstack.backbone.step.ConsentVisualStep;
import org.researchstack.backbone.step.FormStep;
import org.researchstack.backbone.step.InstructionStep;
import org.researchstack.backbone.step.QuestionStep;
import org.researchstack.backbone.task.OrderedTask;
import org.researchstack.backbone.task.Task;
import org.researchstack.backbone.ui.PinCodeActivity;
import org.researchstack.backbone.ui.ViewTaskActivity;
import org.researchstack.backbone.ui.step.layout.ConsentSignatureStepLayout;

import java.util.ArrayList;
import java.util.Collections;

public class MainActivity extends PinCodeActivity
{
    // Activity Request Codes
    private static final int REQUEST_CONSENT = 0;
    private static final int REQUEST_SURVEY  = 1;

    // Task/Step Identifiers
    public static final  String FORM_STEP                 = "form_step";
    public static final  String AGE                       = "age";
    public static final  String INSTRUCTION               = "identifier";
    public static final  String BASIC_INFO_HEADER         = "basic_info_header";
    public static final  String FORM_AGE                  = "form_age";
    public static final  String FORM_GENDER               = "gender";
    public static final  String FORM_MULTI_CHOICE         = "multi_choice";
    public static final  String FORM_DATE_OF_BIRTH        = "date_of_birth";
    public static final  String NUTRITION                 = "nutrition";
    public static final  String SIGNATURE                 = "signature";
    public static final  String SIGNATURE_DATE            = "signature_date";
    public static final  String VISUAL_CONSENT_IDENTIFIER = "visual_consent_identifier";
    public static final  String CONSENT_DOC               = "consent_doc";
    public static final  String SIGNATURE_FORM_STEP       = "form_step";
    public static final  String NAME                      = "name";
    public static final  String CONSENT                   = "consent";
    public static final  String MULTI_STEP                = "multi_step";
    public static final  String DATE                      = "date";
    public static final  String DECIMAL                   = "decimal";
    private static final String FORM_NAME                 = "form_name";
    public static final  String SAMPLE_SURVEY             = "sample_survey";

    // Views
    private AppCompatButton consentButton;
    private AppCompatButton surveyButton;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayShowTitleEnabled(true);

        consentButton = (AppCompatButton) findViewById(R.id.consent_button);
        consentButton.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View v)
            {
                launchConsent();
            }
        });

        surveyButton = (AppCompatButton) findViewById(R.id.survey_button);
        surveyButton.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View v)
            {
                launchSurvey();
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        getMenuInflater().inflate(R.menu.activity_main, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        if (item.getItemId() == R.id.menu_clear)
        {
            clearData();
            Toast.makeText(this, R.string.menu_data_cleared, Toast.LENGTH_SHORT).show();
            return true;
        }
        else
        {
            return super.onOptionsItemSelected(item);
        }
    }

    void clearData()
    {
        AppPrefs appPrefs = AppPrefs.getInstance(this);
        appPrefs.setHasSurveyed(false);
        appPrefs.setHasConsented(false);

        initViews();
    }

    @Override
    public void onDataReady()
    {
        super.onDataReady();
        initViews();
    }

    private void initViews()
    {
        AppPrefs prefs = AppPrefs.getInstance(this);

        View lblConsentedDate = findViewById(R.id.consented_date_lbl);
        TextView consentedDate = (TextView)findViewById(R.id.consented_date);
        ImageView consentedSig = (ImageView) findViewById(R.id.consented_signature);

        if(prefs.hasConsented())
        {
            consentButton.setVisibility(View.GONE);
            surveyButton.setEnabled(true);

            consentedSig.setVisibility(View.VISIBLE);
            consentedDate.setVisibility(View.VISIBLE);
            lblConsentedDate.setVisibility(View.VISIBLE);

            printConsentInfo(consentedDate, consentedSig);
        }
        else
        {
            consentButton.setVisibility(View.VISIBLE);
            surveyButton.setEnabled(false);

            consentedSig.setVisibility(View.INVISIBLE);
            consentedSig.setImageBitmap(null);
            consentedDate.setVisibility(View.INVISIBLE);
            lblConsentedDate.setVisibility(View.INVISIBLE);
        }

        TextView surveyAnswer = (TextView) findViewById(R.id.survey_results);

        if(prefs.hasSurveyed())
        {
            surveyAnswer.setVisibility(View.VISIBLE);
            printSurveyInfo(surveyAnswer);
        }
        else
        {
            surveyAnswer.setVisibility(View.GONE);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data)
    {
        super.onActivityResult(requestCode, resultCode, data);

        if(requestCode == REQUEST_CONSENT && resultCode == RESULT_OK)
        {
            processConsentResult((TaskResult) data.getSerializableExtra(ViewTaskActivity.EXTRA_TASK_RESULT));
        }
        else if(requestCode == REQUEST_SURVEY && resultCode == RESULT_OK)
        {
            processSurveyResult((TaskResult) data.getSerializableExtra(ViewTaskActivity.EXTRA_TASK_RESULT));
        }
    }

    // Consent stuff

    private void launchConsent()
    {
        ConsentDocument document = new ConsentDocument();
        document.setTitle("Demo Consent");
        document.setSignaturePageTitle(R.string.rsb_consent);

        // Create consent visual sections
        ConsentSection section1 = new ConsentSection(ConsentSection.Type.DataGathering);
        section1.setTitle("The title of the section goes here ...");
        section1.setSummary("The summary about the section goes here ...");
        section1.setContent("The content to show in learn more ...");

        // ...add more sections as needed, then create a visual consent step
        ConsentVisualStep visualStep = new ConsentVisualStep(VISUAL_CONSENT_IDENTIFIER);
        visualStep.setStepTitle(R.string.rsb_consent);
        visualStep.setSection(section1);
        visualStep.setNextButtonString(getString(R.string.rsb_next));

        // Create consent signature object and set what info is required
        ConsentSignature signature = new ConsentSignature();
        signature.setRequiresName(true);
        signature.setRequiresSignatureImage(true);

        // Create our HTML to show the user and have them accept or decline.
        StringBuilder docBuilder = new StringBuilder(
                "</br><div style=\"padding: 10px 10px 10px 10px;\" class='header'>");
        String title = getString(R.string.rsb_consent_review_title);
        docBuilder.append(String.format(
                "<h1 style=\"text-align: center; font-family:sans-serif-light;\">%1$s</h1>",
                title));
        String detail = getString(R.string.rsb_consent_review_instruction);
        docBuilder.append(String.format("<p style=\"text-align: center\">%1$s</p>", detail));
        docBuilder.append("</div></br>");
        docBuilder.append("<div><h2> HTML Consent Doc goes here </h2></div>");

        // Create the Consent doc step, pass in our HTML doc
        ConsentDocumentStep documentStep = new ConsentDocumentStep(CONSENT_DOC);
        documentStep.setConsentHTML(docBuilder.toString());
        documentStep.setConfirmMessage(getString(R.string.rsb_consent_review_reason));

        // Create Consent form step, to get users first & last name
        FormStep formStep = new FormStep(SIGNATURE_FORM_STEP,
                "Form Title",
                "Form step description");
        formStep.setStepTitle(R.string.rsb_consent);

        TextAnswerFormat format = new TextAnswerFormat();
        format.setIsMultipleLines(false);

        QuestionStep fullName = new QuestionStep(NAME, "Full name", format);
        formStep.setFormSteps(Collections.singletonList(fullName));

        // Create Consent signature step, user can sign their name
        ConsentSignatureStep signatureStep = new ConsentSignatureStep(SIGNATURE);
        signatureStep.setStepTitle(R.string.rsb_consent);
        signatureStep.setTitle(getString(R.string.rsb_consent_signature_title));
        signatureStep.setText(getString(R.string.rsb_consent_signature_instruction));
        signatureStep.setSignatureDateFormat(signature.getSignatureDateFormatString());
        signatureStep.setOptional(false);
        signatureStep.setStepLayoutClass(ConsentSignatureStepLayout.class);

        // Finally, create and present a task including these steps.
        Task consentTask = new OrderedTask(CONSENT,
                visualStep,
                documentStep,
                formStep,
                signatureStep);

        // Launch using ViewBackboneInteropTaskActivity and make sure to listen for the activity result
        Intent intent = ViewBackboneInteropTaskActivity.newIntent(this, consentTask);
        startActivityForResult(intent, REQUEST_CONSENT);
    }

    private void processConsentResult(TaskResult result)
    {
        boolean consented = (boolean) result.getStepResult(CONSENT_DOC).getResult();

        if(consented)
        {
            StorageAccess.getInstance().getAppDatabase().saveTaskResult(result);

            AppPrefs prefs = AppPrefs.getInstance(this);
            prefs.setHasConsented(true);

            initViews();
        }
    }

    private void printConsentInfo(TextView consentedDate, ImageView consentedSig)
    {
        TaskResult result = StorageAccess.getInstance()
                .getAppDatabase()
                .loadLatestTaskResult(CONSENT);

        String signatureBase64 = (String) result.getStepResult(SIGNATURE)
                .getResultForIdentifier(ConsentSignatureStepLayout.KEY_SIGNATURE);

        String signatureDate = (String) result.getStepResult(SIGNATURE)
                .getResultForIdentifier(ConsentSignatureStepLayout.KEY_SIGNATURE_DATE);

        consentedDate.setText(signatureDate);

        byte[] signatureBytes = Base64.decode(signatureBase64, Base64.DEFAULT);
        consentedSig.setImageBitmap(BitmapFactory.decodeByteArray(
                signatureBytes,
                0,
                signatureBytes.length));
    }


    // Survey Stuff

    private void launchSurvey()
    {
        InstructionStep instructionStep = new InstructionStep(INSTRUCTION,
                "Selection Survey",
                "This survey can help us understand your eligibility for the fitness study");
        instructionStep.setStepTitle(R.string.survey);

        TextAnswerFormat format = new TextAnswerFormat();
        QuestionStep ageStep = new QuestionStep(NAME, "What is your name?", format);
        ageStep.setStepTitle(R.string.survey);

        DateAnswerFormat dateFormat = new DateAnswerFormat(AnswerFormat.DateAnswerStyle.Date);
        QuestionStep dateStep = new QuestionStep(DATE, "Enter a date", dateFormat);
        dateStep.setStepTitle(R.string.survey);

        // Create a Boolean step to include in the task.
        QuestionStep booleanStep = new QuestionStep(NUTRITION);
        booleanStep.setStepTitle(R.string.survey);
        booleanStep.setTitle("Do you take nutritional supplements?");
        booleanStep.setAnswerFormat(new BooleanAnswerFormat(getString(R.string.rsb_yes),
                getString(R.string.rsb_no)));
        booleanStep.setOptional(false);

        QuestionStep multiStep = new QuestionStep(MULTI_STEP);
        multiStep.setStepTitle(R.string.survey);
        AnswerFormat multiFormat = new ChoiceAnswerFormat(AnswerFormat.ChoiceAnswerStyle.MultipleChoice,
                new Choice<>("Zero", 0),
                new Choice<>("One", 1),
                new Choice<>("Two", 2));
        multiStep.setTitle("Select multiple");
        multiStep.setAnswerFormat(multiFormat);
        multiStep.setOptional(false);

        // Create a task wrapping the steps.
        OrderedTask task = new OrderedTask(SAMPLE_SURVEY, instructionStep, ageStep, dateStep,
                // formStep,
                booleanStep, multiStep);

        // Create an activity using the task and set a delegate.
        Intent intent = ViewBackboneInteropTaskActivity.newIntent(this, task);
        startActivityForResult(intent, REQUEST_SURVEY);
    }

    @NonNull
    private FormStep createFormStep()
    {
        FormStep formStep = new FormStep(FORM_STEP, "Form", "Form groups multi-entry in one page");
        ArrayList<QuestionStep> formItems = new ArrayList<>();

        QuestionStep basicInfoHeader = new QuestionStep(BASIC_INFO_HEADER,
                "Basic Information",
                new UnknownAnswerFormat());
        formItems.add(basicInfoHeader);

        TextAnswerFormat format = new TextAnswerFormat();
        format.setIsMultipleLines(false);
        QuestionStep nameItem = new QuestionStep(FORM_NAME, "Name", format);
        formItems.add(nameItem);

        QuestionStep ageItem = new QuestionStep(FORM_AGE, "Age", new IntegerAnswerFormat(18, 90));
        formItems.add(ageItem);

        AnswerFormat genderFormat = new ChoiceAnswerFormat(AnswerFormat.ChoiceAnswerStyle.SingleChoice,
                new Choice<>("Male", 0),
                new Choice<>("Female", 1));
        QuestionStep genderFormItem = new QuestionStep(FORM_GENDER, "Gender", genderFormat);
        formItems.add(genderFormItem);

        AnswerFormat multiFormat = new ChoiceAnswerFormat(AnswerFormat.ChoiceAnswerStyle.MultipleChoice,
                new Choice<>("Zero", 0),
                new Choice<>("One", 1),
                new Choice<>("Two", 2));
        QuestionStep multiFormItem = new QuestionStep(FORM_MULTI_CHOICE, "Test Multi", multiFormat);
        formItems.add(multiFormItem);

        AnswerFormat dateOfBirthFormat = new DateAnswerFormat(AnswerFormat.DateAnswerStyle.Date);
        QuestionStep dateOfBirthFormItem = new QuestionStep(FORM_DATE_OF_BIRTH,
                "Birthdate",
                dateOfBirthFormat);
        formItems.add(dateOfBirthFormItem);

        // ... And so on, adding additional items
        formStep.setFormSteps(formItems);
        return formStep;
    }

    private void processSurveyResult(TaskResult result)
    {
        StorageAccess.getInstance().getAppDatabase().saveTaskResult(result);

        AppPrefs prefs = AppPrefs.getInstance(this);
        prefs.setHasSurveyed(true);
        initViews();
    }

    private void printSurveyInfo(TextView surveyAnswer)
    {
        TaskResult taskResult = StorageAccess.getInstance()
                .getAppDatabase()
                .loadLatestTaskResult(SAMPLE_SURVEY);

        String results = "";
        for(String id : taskResult.getResults().keySet())
        {
            StepResult stepResult = taskResult.getStepResult(id);
            results += id + ": " + stepResult.getResult() + "\n";
        }

        surveyAnswer.setText(results);
    }
}

File: foundationexampleapp/src/main/java/org/researchstack/foundationexampleapp/MainActivity.kt
package org.researchstack.kotlinbackbonesampleapp


import android.content.Context
import android.graphics.BitmapFactory
import android.os.Bundle
import android.util.Base64
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import androidx.appcompat.widget.Toolbar
import androidx.core.os.CancellationSignal
import org.researchstack.feature.authentication.pincode.PasscodeAuthenticator
import org.researchstack.feature.authentication.pincode.ui.PasscodeAuthenticationFragment
import org.researchstack.feature.storage.StorageAccess
import org.researchstack.feature.storage.file.StorageAccessListener
import org.researchstack.feature.consent.ui.layout.ConsentSignatureStepLayout
import org.researchstack.foundation.components.presentation.TaskPresentationCallback
import org.researchstack.foundation.components.presentation.compatibility.BackwardsCompatibleStepFragmentProvider
import org.researchstack.foundation.components.presentation.compatibility.BackwardsCompatibleStepLayoutProvider
import org.researchstack.foundation.components.presentation.compatibility.BackwardsCompatibleTaskPresentationFragment
import org.researchstack.foundation.components.utils.LogExt
import org.researchstack.foundation.core.models.result.TaskResult
import org.researchstack.foundation.core.models.task.Task
import org.researchstack.foundationexampleapp.R

class MainActivity : AppCompatActivity(), StorageAccessListener, PasscodeAuthenticator.PresentationDelegate {

    companion object {
        val TAG = MainActivity.javaClass.name
    }

    // Views
    private var consentButton: AppCompatButton? = null
    private var surveyButton: AppCompatButton? = null
    private var taskPresentationFragment: BackwardsCompatibleTaskPresentationFragment? = null

    val taskProvider = FoundationTaskProvider(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        this.supportActionBar!!.setDisplayShowTitleEnabled(true)

        this.consentButton = findViewById(R.id.consent_button) as AppCompatButton
        this.consentButton!!.setOnClickListener {
            launchConsent()
        }

        this.surveyButton = findViewById(R.id.survey_button) as AppCompatButton
        this.surveyButton!!.setOnClickListener {
            launchSurvey()
        }

        val passcodeAuthenticator: PasscodeAuthenticator = FoundationApplication.instance!!.passcodeAuthenticator!!
        passcodeAuthenticator.setPresentationDelegate(this)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_main, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_clear) {
            clearData()
            Toast.makeText(this, R.string.menu_data_cleared, Toast.LENGTH_SHORT).show()
            return true
        }
        else if (item.itemId == android.R.id.home) {
            this.taskPresentationFragment?.onBackPressed()
            return true
        }
        else {
            return super.onOptionsItemSelected(item)
        }
    }

    private fun clearData() {
        val appPrefs = AppPrefs.getInstance(this)
        appPrefs.setHasSurveyed(false)
        appPrefs.setHasConsented(false)

        initViews()
    }

    override fun onDataReady() {
        this.storageAccessUnregister()
        initViews()
    }


    override fun onDataAuth() {

    }

    override fun presentPasscodeAuthentication(authenticator: PasscodeAuthenticator, cancel: CancellationSignal, callback: PasscodeAuthenticator.PasscodeAuthenticationCallback) {
//        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.

        //present pin code fragment

        val context = this as Context

        val fragmentCallback = object: PasscodeAuthenticationFragment.AuthenticationCallback() {
            override fun onAuthenticationFailed() {
                callback.onAuthenticationFailed()
            }

            override fun onAuthenticationSucceeded() {
                callback.onAuthenticationSucceeded(PasscodeAuthenticator.PasscodeAuthenticationResult())

                supportFragmentManager.findFragmentByTag("passcode_fragment")?.let { fragment ->
                    supportFragmentManager.beginTransaction()
                            .setCustomAnimations(R.anim.rsf_slide_in_up, R.anim.rsf_slide_out_down)
                            .remove(fragment)
                            .commit()
                }

                //need to request storage access again here!!
                StorageAccess.getInstance().requestStorageAccess(context)
            }
        }

        val passcodeFragment = PasscodeAuthenticationFragment.newInstance(authenticator, fragmentCallback)
        val transaction = supportFragmentManager.beginTransaction()
                .replace(R.id.authorization_container, passcodeFragment, "passcode_fragment")
                .commit()

        supportFragmentManager.executePendingTransactions()
    }

    override fun onDataFailed() {
        this.storageAccessUnregister()
    }

    fun requestStorageAccess() {
        LogExt.i(TAG, "requestStorageAccess()")
        val storageAccess = StorageAccess.getInstance()
        storageAccessRegister()
        storageAccess.requestStorageAccess(this)
    }

    fun storageAccessRegister() {
        LogExt.i(TAG, "storageAccessRegister()")
        val storageAccess = StorageAccess.getInstance()
        storageAccess.register(this)
    }

    fun storageAccessUnregister() {
        LogExt.i(TAG, "storageAccessUnregister()")
        val storageAccess = StorageAccess.getInstance()
        storageAccess.unregister(this)
    }

    override fun onResume() {
        super.onResume()

        val storageAccess = StorageAccess.getInstance()
        if (!storageAccess.hasPinCode(this)) {
            this.launchPinCodeRegistration()
        }
        else {
            this.requestStorageAccess()
        }

    }

    private fun launchTask(taskIdentifier:String) {

        val layoutProvider = BackwardsCompatibleStepLayoutProvider()
        val stepFragmentProvider = BackwardsCompatibleStepFragmentProvider(layoutProvider)

        val callback: TaskPresentationCallback<TaskResult, Task> = object: TaskPresentationCallback<TaskResult, Task>() {
            override fun onTaskPresentationFinished(task: Task, result: TaskResult?) {

                taskPresentationFragment = null
                supportFragmentManager.findFragmentByTag(taskIdentifier)?.let { fragment ->
                    supportFragmentManager.beginTransaction()
                            .setCustomAnimations(R.anim.rsf_slide_in_up, R.anim.rsf_slide_out_down)
                            .remove(fragment)
                            .commit()
                }

                result?.let {
                    processTaskResult(task, it)
                }

            }
        }

        val taskPresentationFragment = BackwardsCompatibleTaskPresentationFragment.newInstance(taskIdentifier, stepFragmentProvider, callback)
        taskPresentationFragment.taskProvider = this.taskProvider

        this.taskPresentationFragment = taskPresentationFragment

        supportFragmentManager.beginTransaction()
                .setCustomAnimations(R.anim.rsf_slide_in_up, R.anim.rsf_slide_out_down)
                .add(R.id.task_presentation_container, taskPresentationFragment, taskIdentifier)
                .commit()

    }

    private fun processTaskResult(task: Task, result: TaskResult) {
        if (task.identifier == FoundationTaskProvider.PIN_CODE) {
            this.processPinCodeResult(result)
        }
        else if (task.identifier == FoundationTaskProvider.CONSENT) {
            this.processConsentResult(result)
        }
        else if (task.identifier == FoundationTaskProvider.SAMPLE_SURVEY) {
            this.processSurveyResult(result)
        }
    }

    private fun launchPinCodeRegistration() {
        this.launchTask(FoundationTaskProvider.PIN_CODE)
    }

    private fun processPinCodeResult(result: TaskResult) {
        val pinCode: String = result.getStepResult(FoundationTaskProvider.PIN_CODE)!!.result as String
        val storageAccess = StorageAccess.getInstance()
        storageAccess.createPinCode(this, pinCode)
    }

    private fun initViews() {
        val prefs = AppPrefs.getInstance(this)

        val lblConsentedDate = findViewById(R.id.consented_date_lbl) as TextView
        val consentedDate = findViewById(R.id.consented_date) as TextView
        val consentedSig = findViewById(R.id.consented_signature) as ImageView

        if (prefs.hasConsented()) {
            this.consentButton!!.setVisibility(View.GONE)
            this.surveyButton!!.setEnabled(true)

            consentedSig.visibility = View.VISIBLE
            consentedDate.visibility = View.VISIBLE
            lblConsentedDate.setVisibility(View.VISIBLE)

            printConsentInfo(consentedDate, consentedSig)
        } else {
            this.consentButton!!.setVisibility(View.VISIBLE)
            this.surveyButton!!.setEnabled(false)

            consentedSig.visibility = View.INVISIBLE
            consentedSig.setImageBitmap(null)
            consentedDate.visibility = View.INVISIBLE
            lblConsentedDate.setVisibility(View.INVISIBLE)
        }

        val surveyAnswer = findViewById(R.id.survey_results) as TextView

        if (prefs.hasSurveyed()) {
            surveyAnswer.visibility = View.VISIBLE
            printSurveyInfo(surveyAnswer)
        } else {
            surveyAnswer.visibility = View.GONE
        }
    }

    // Consent stuff

    private fun launchConsent() {
        this.launchTask(FoundationTaskProvider.CONSENT)
    }

    private fun processConsentResult(result: TaskResult) {
        val consented = result.getStepResult(FoundationTaskProvider.CONSENT_DOC).result as Boolean

        if (consented) {
            StorageAccess.getInstance().appDatabase.saveTaskResult(result)

            val prefs = AppPrefs.getInstance(this)
            prefs.setHasConsented(true)

            initViews()
        }
    }

    private fun printConsentInfo(consentedDate: TextView, consentedSig: ImageView) {
        val result = StorageAccess.getInstance()
                .appDatabase
                .loadLatestTaskResult(FoundationTaskProvider.CONSENT)

        val signatureBase64 = result.getStepResult(FoundationTaskProvider.SIGNATURE)
                .getResultForIdentifier(ConsentSignatureStepLayout.KEY_SIGNATURE) as String

        val signatureDate = result.getStepResult(FoundationTaskProvider.SIGNATURE)
                .getResultForIdentifier(ConsentSignatureStepLayout.KEY_SIGNATURE_DATE) as String

        consentedDate.text = signatureDate

        val signatureBytes = Base64.decode(signatureBase64, Base64.DEFAULT)
        consentedSig.setImageBitmap(BitmapFactory.decodeByteArray(
                signatureBytes,
                0,
                signatureBytes.size))
    }


    // Survey Stuff

    private fun launchSurvey() {
        this.launchTask(FoundationTaskProvider.SAMPLE_SURVEY)
    }

    private fun processSurveyResult(result: TaskResult) {
        StorageAccess.getInstance().appDatabase.saveTaskResult(result)

        val prefs = AppPrefs.getInstance(this)
        prefs.setHasSurveyed(true)
        initViews()
    }

    private fun printSurveyInfo(surveyAnswer: TextView) {
        val taskResult = StorageAccess.getInstance()
                .appDatabase
                .loadLatestTaskResult(FoundationTaskProvider.SAMPLE_SURVEY)

        var results = ""
        for (id in taskResult.results.keys) {
            val stepResult = taskResult.getStepResult(id)
            results += id + ": " + stepResult.result.toString() + "\n"
        }

        surveyAnswer.text = results
    }



}
